### PR TITLE
Fix complete attribute calculation for backward pagination in MAM

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR is rebased #1541.

### CODE REVIEW

File: src/mod_mam_utils.erl [hunk 1]
```diff
@@ -969,11 +969,22 @@ maybe_previous_id(X) ->
     X - 1.
 
 
--spec is_last_page(PageSize, TotalCount, Offset, MessageRows) -> boolean() when
+%% @doc Returns true, if there are more pages to paginate in the result set.
```
Comment (kzemek) [inline src/mod_mam_utils.erl]: This is why I disagree with over-commenting - the function is called `is_complete_result` which suggests that it returns `true` if this is the final page of the result. The comment says something opposite though. The function name is right while the comment is already outdated on introduction.

Comment (arcusfelis) [reply src/mod_mam_utils.erl]: yep, should be "if there are no more pages"...

Comment (arcusfelis) [reply src/mod_mam_utils.erl]: fixed the comment.
Also added some more info, about future optimizations (the one, that my client uses).

File: test.disabled/ejabberd_tests/tests/mam_helper.erl [hunk 2]
```diff
@@ -156,6 +158,14 @@ assert_respond_query_id(_P, _ExpectedQueryId, #result_iq{query_id=not_supported}
 assert_respond_query_id(_P, ExpectedQueryId, #result_iq{query_id=QueryId}) ->
     ?assert_equal(ExpectedQueryId, QueryId).
 
+%% @doc Calls wait_archive_respond and checks "complete" field value
+wait_for_complete_respond(P, Alice, Expected) ->
```
Comment (fenek) [inline test.disabled/ejabberd_tests/tests/mam_helper.erl]: 1. `Expected` is not very obvious on **what** is expected. Proper name would even obsolete the comment.
2. Perhaps `wait_for_complete_response`?

File: test.disabled/ejabberd_tests/tests/mam_SUITE.erl [hunk 3]
```diff
@@ -2538,6 +2571,191 @@ pagination_last_after_id5_before_id11(Config) ->
...
+%% Less than page_size are returned.
+%% GIVEN 15 archived messages
+%% WHEN direction=after, rsm_id=10, page_size=5
+%% THAN complete=true
+after_complete_true_after11(Config) ->
+    P = ?config(props, Config),
+    F = fun(Alice) ->
+        %% Get the last page of size 5.
+        %% Get messages: 12,13,14,15
+        RSM = #rsm_in{max=5, direction='after', id=message_id(10, Config)},
```
Comment (fenek) [inline test.disabled/ejabberd_tests/tests/mam_SUITE.erl]: Shouldn't it be `message_id(11, Config)`?

File: test.disabled/ejabberd_tests/tests/mam_SUITE.erl [hunk 4]
```diff
@@ -2538,6 +2571,191 @@ pagination_last_after_id5_before_id11(Config) ->
...
+    F = fun(Alice) ->
+        %% Get messages: 1,2,3,4,5
+        RSM = #rsm_in{max=5, direction=before, id=message_id(6, Config)},
+        rsm_send(Config, Alice,
+            stanza_page_archive_request(P, <<"before6">>, RSM)),
+        wait_for_complete_respond(P, Alice, <<"true">>)
+        end,
+    parallel_story(Config, [{alice, 1}], F).
+
+%% First page is not compete, because max is smaller than archive size.
```
Comment (fenek) [inline test.disabled/ejabberd_tests/tests/mam_SUITE.erl]: `complete`

File: test.disabled/ejabberd_tests/tests/mam_SUITE.erl [hunk 5]
```diff
@@ -2538,6 +2571,191 @@ pagination_last_after_id5_before_id11(Config) ->
...
+            stanza_page_archive_request(P, <<"after10">>, RSM)),
+        wait_for_complete_respond(P, Alice, <<"true">>)
+        end,
+    parallel_story(Config, [{alice, 1}], F).
+
+%% There are no messages to paginate after this request.
+%% Less than page_size are returned.
+%% GIVEN 15 archived messages
+%% WHEN direction=after, rsm_id=10, page_size=5
+%% THAN complete=true
```
Comment (fenek) [inline test.disabled/ejabberd_tests/tests/mam_SUITE.erl]: `THAN` => `THEN` in all comments in new test cases.

File: test.disabled/ejabberd_tests/tests/mam_SUITE.erl [hunk 6]
```diff
@@ -2538,6 +2571,191 @@ pagination_last_after_id5_before_id11(Config) ->
...
+            stanza_page_archive_request(P, <<"before1">>, RSM)),
+        wait_for_complete_respond(P, Alice, <<"true">>)
+        end,
+    parallel_story(Config, [{alice, 1}], F).
+
+%% Reaches the end of result set
+%% GIVEN 15 archived messages
+%% WHEN direction=before, rsm_id=2, page_size=5
+%% THAN complete=true
+before_complete_true_before2(Config) ->
```
Comment (fenek) [inline test.disabled/ejabberd_tests/tests/mam_SUITE.erl]: Is this test case that much different from `before_complete_true_before5`? Does it test different branch of execution?

File: src/mod_mam_utils.erl [hunk 7]
```diff
@@ -969,23 +969,58 @@ maybe_previous_id(X) ->
...
+%%
+%% TotalCount and Offset can be undefined, in case we use IsSimple=true.
+%% IsSimple=true tells the server not to do heavy `SELECT COUNT(*)' queries.
+%%
+%% TODO It is possible to set complete flag WITH IsSimple=true,
+%%      if we select one extra message from archive, but don't send it to the client.
+%%      It's the most efficient way to query archive, if the client side does
+%%      not care about the total number of messages and if it's stateless
+%%      (i.e. web interface).
+-spec is_complete_result_page(PageSize, TotalCount, Offset, MessageRows, Params) ->
```
Comment (fenek) [inline src/mod_mam_utils.erl]: I think we can assume that `PageSize` is a part of `Params`?

File: test.disabled/ejabberd_tests/tests/mam_SUITE.erl (hunk 5 above)
Comment (arcusfelis) [reply test.disabled/ejabberd_tests/tests/mam_SUITE.erl]: yeah

Comment (fenek) [reply test.disabled/ejabberd_tests/tests/mam_SUITE.erl]: Why?

File: src/mod_mam_utils.erl (hunk 7 above)
Comment (arcusfelis) [reply src/mod_mam_utils.erl]: Unless it's get min(PageSizeFromClient, MaxPageSizeInOpts).
Need to check.

File: test.disabled/ejabberd_tests/tests/mam_SUITE.erl (hunk 6 above)
Comment (arcusfelis) [reply test.disabled/ejabberd_tests/tests/mam_SUITE.erl]: I will remove this test.

**Comment by arcusfelis** [comment]
Weird. For some reason, clients in rsm tests don't receive a presence from a server. 
http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1740/4182/riak_mnesia.18.3/big/index.html

Maybe because they all use the same username to test archive in parallel, and some extra test cases in the group overloaded MongooseIM (that should NOT happen).

I will restart the travis job to look if it's just a glitch.

**Comment by arcusfelis** [comment]
still broken even after reducing concurrency
http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1740/4223/riak_mnesia.18.3/big/ct_run.test@travis-job-c6222ab1-a645-4a78-ba70-0d58e8b73aeb.2018-03-03_10.01.21/ejabberd_tests.tests.mam_SUITE.logs/run.2018-03-03_10.09.17/suite.log.html

**Comment by arcusfelis** [comment]
Because codecov doesn't like force pushes.

[Last commit](https://codecov.io/gh/esl/MongooseIM/commit/54af7e603adac26b89e39ae00bef878a7195c3e8)
[Diff `master..this_branch`](https://codecov.io/gh/esl/MongooseIM/compare/75c8d6c6cfc89279445f42745c5c6a58e630c2dd...54af7e603adac26b89e39ae00bef878a7195c3e8)

**Pull request merged**